File: src/Codes/C++/quicksort.cpp
#include <iostream>

// Function to partition the array and return the pivot index
int partition(int *arr, int start, int end)
{
    int pivot = arr[end]; // Choose the last element as the pivot
    int i = start - 1;   // Index of the smaller element
 
    // Loop through the array and swap elements if necessary
    for (int j = start; j < end; j++)
    {
        // If the current element is smaller than the pivot
        if (arr[j] < pivot)
        {
            i++; // Increment the index of the smaller element
            std::swap(arr[i], arr[j]); // Swap the elements
        }
    }
 
    std::swap(arr[i + 1], arr[end]); // Swap the pivot element into its correct position
    return i + 1; // Return the pivot index
}

// Recursive quicksort function
void quicksort(int *arr, int start, int end)
{
    if (start < end) // If the array has more than one element
    {
        int pivotIndex = partition(arr, start, end); // Partition the array and get the pivot index
 
        // Sort the elements before and after the pivot
        quicksort(arr, start, pivotIndex - 1);
        quicksort(arr, pivotIndex + 1, end);
    }
}

int main()
{
    int arr[] = {5, 3, 6, 2, 10};
    int n = sizeof(arr) / sizeof(arr[0]);
 
    // Sort the array
    quicksort(arr, 0, n - 1);
 
    // Print the sorted array
    for (int i = 0; i < n; i++)
        std::cout << arr[i] << " ";
 
    return 0;
}

File: src/Codes/Python/quicksort.py
def partition(arr, start, end):
    # Choose the last element as the pivot
    pivot = arr[end]
 
    # Index of the smaller element
    i = start - 1
 
    # Loop through the array and swap elements if necessary
    for j in range(start, end):
        # If the current element is smaller than the pivot
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
 
    # Swap the pivot element into its correct position
    arr[i + 1], arr[end] = arr[end], arr[i + 1]
 
    return i + 1

def quicksort(arr, start, end):
    if start < end:
        # Partition the array and get the pivot index
        pivot_index = partition(arr, start, end)
 
        # Sort the elements before and after the pivot
        quicksort(arr, start, pivot_index - 1)
        quicksort(arr, pivot_index + 1, end)

# Test the function
print(quicksort([5, 3, 6, 2, 10], 0, 4)) # [2, 3, 5, 6, 10]
